### DISCUSSION TITLE
AEP - cut in wind speed higher than some values in wind rose data

### DISCUSSION THREAD

**Discussion by Nellytado** [NatLabRockies/floris#863] [Q&A] [answered]
Hi, when calculating the AEP for a farm, if we have our own wind rose data with a list of wind speeds, wind directions and corresponding freq values, and the turbine file cut-in speed is 4 m/s (V80 turbine) how can we accurately calculate AEP if the wind rose data has speeds below this? 

I get this error: ValueError: A value (0.39905441321260154) in x_new is below the interpolation range's minimum value (4.0). 

Should I remove all wind speeds below 4m/s in my wind rose data? Would this not make the freq val which have to sum up to one inaccurate? If anyone knows how to get past this problem would be much appreciated. 

Thank you,

Undergraduate Student writing final year project.

**Comment by misi9170** [comment]
Hi @Nellytado ,

I'm sorry for the slow reply. The formula for calculating AEP can be written as 

$$ \text{AEP} = \sum_i P(i) p(i) \times \text{HPY} $$

where $i$ is the condition to be evaluate (usually, a (wind speed, wind direction) pair); $P(i)$ is the power produced by the wind farm in that condition; $p(i)$ is the frequency of occurrence of that condition in the wind rose; and $\text{HPY}$ is the hours in a year, usually taken as $365 \times 24 = 8760$. If power is reported in Watts, the resulting annual energy production has units Watt-hours. This can alternatively be written

$$ \text{AEP} = \texttt{E}[P] \times \text{HPY}$$

where $\texttt{E}[P]$ is the expected power of the wind farm across all wind rose conditions.

In your case, a condition with a wind speed below the cut-in speed for the turbine (say, 3m/s) produces zero power. This will be multiplied by a _non-zero_ probability of that wind speed occurring, and the product $P(i) p(i)$ will be zero; but that is still correct.

So, you may proceed to remove the cases below 4m/s from the wind rose. Your frequency values will no longer sum to one, but that is OK; you are effectively "skipping" those terms in the sum, as you already know that their contribution to the AEP is zero. You shouldn't redistribute the frequency assigned to cases below 4m/s, as that will erroneously upweight the AEP. 

However, this solution won't fix a potential issue that when the freestream wind speed is just above the cut-in speed (say, 5m/s), but some turbines are waked and so their inflow is lower than 4m/s, the error you are seeing may still arise. A better solution would be to alter the turbine input file for the V80 that you have to specify that at low wind speeds, the power (or Cp, if you are using FLORIS v3) is zero; that is, add a few entries to the list of wind speeds, powers, and thrust coefficients on the turbine input file that specify zero power below cut in. This will avoid the interpolation error you are seeing and let you calculate the "complete" wind rose, including the zero terms, explicitly.

**Reply by Nellytado** [reply]
Hi, thank you this was explained very well and will attempt as suggested! That means we can also calculate the AEP if we get turbine powers and multiply it by 8760 yes? 

This is the line of code I used in the sweep wind directions example so I can show AEP against wind direction 'turbine_powers = fi.get_turbine_powers()*8760'

**Reply by misi9170** [reply]
Yes, if we let $P_t(i)$ be the power of turbine $t$ in condition $i$, so that the farm power $P(i)$ can be expressed $P(i) = \sum_t P_t(i)$, then the AEP calculation can alternatively be written as 

$$ \text{AEP} = \sum_i \left[ \sum_t P_t(i) \right] p(i) \times \text{HPY} $$

or, by linearity,

$$ \text{AEP} = \sum_i \left[ \sum_t P_t(i) \times \text{HPY} \right] p(i) $$

If I'm understanding what you'd like to do correctly, then `fi.get_turbine_powers()*8760` is equivalent to the $P_t(i) \times \text{HPY}$ term in the equation above. 

You'll need to be a bit careful about what you call this. "AEP against wind direction" is unclear, because by the usual definition, AEP includes all wind directions seen in a year. The product `fi.get_turbine_powers()*8760` will give you the power of each turbine in each condition multiplied by the hours in a year. If these are multipled by the frequency of occurrence for each condition, this may be considered the "portion" of AEP provided by each turbine in each condition.

**Reply by Nellytado** [reply]
Thank you, this was super helpful! I appreciate the time taken to help me understand.

**Reply by Nellytado** [reply]
I calculate my 'portion of AEP for wind speed 8m/s' using the following code on FLORIS 3.6. I have a 30 turbine farm using vest V80 turbines (cc input file attached and V80 turbine file attached). However, the graph it generates(below) cannot be correct as using V80 turbine files (max power is 2MW) the total farm power should generate in a year on average 2MW x 30 x 8760 x 0.11103 = 58357MWh for that wind speed. 

The graph shows a max power of 1.4 x 1E11 MWh which is higher than the predicted 58375.MWh. I do not understand why and this must be incorrect. Any help is much appreciated as this is for a student project and I am very passionate about it!

The graph generated
![image](https://github.com/NREL/floris/assets/165584201/ff456a75-7df5-4ef0-9f17-0d0e18f5a66e)

cc input file 
[ccNH.txt](https://github.com/NREL/floris/files/15069502/ccNH.txt)

V80 turbine file 
[V80.txt](https://github.com/NREL/floris/files/15069467/V80.txt)



`import numpy as np
import matplotlib.pyplot as plt
from floris.tools import FlorisInterface
import matplotlib.ticker as ticker

# Instantiate the FLORIS interface
fi = FlorisInterface("inputs/ccNH.yaml")

# Sweep wind directions from 0 to 360 degrees
wd_array = np.arange(0, 360)
fi.reinitialize(wind_directions=wd_array)


# Define a matrix of yaw angles to be all zeros based on the original layout from the file
num_wd = len(wd_array)  # Number of wind directions
num_ws = 1  # Number of wind speeds (assuming constant wind speed)
num_turbines = fi.floris.farm.n_turbines  # Get the number of turbines from the FLORIS object
yaw_angles = np.zeros((num_wd, num_ws, num_turbines))

# Calculate wake
fi.calculate_wake(yaw_angles=yaw_angles)

# Collect turbine powers and sum them for each wind direction
turbine_powers = fi.get_turbine_powers()*8760*0.11103/1E3 #total farm power multiplied by hours in a year and speed frequency 
total_power = np.sum(turbine_powers, axis=2) # Sum power from turbines

# Plot total power vs. wind direction
fig, ax = plt.subplots()
ax.plot(wd_array, total_power,color='r', label='Farm')


ax.grid(True)
ax.set_xlabel('Wind Direction (deg)')
ax.set_ylabel('Part of Annual Energy Production due to Wind Speed (MWh)')
plt.legend()
plt.show()`

**Reply by misi9170** [reply, marked as answer]
HI @Nellytado , the issue here is the way you've defined power in the turbine input file. For FLORIS v3, the `power` field on the turbine input file should be a power coefficient, with a theoretical upper limit of 16/27. See [here](https://www.ftexploring.com/wind-energy/wind-power-coefficient.htm) for an explanation.

In your turbine input file, power is specified in absolute terms (kW). This is how power is specified in FLORIS v4.0, but not in version 3. Here is an [example turbine input file](https://github.com/NREL/floris/blob/v3/floris/turbine_library/nrel_5MW.yaml) from FLORIS v3, showing how the `power` field should be defined. You will need to convert your power in absolute terms to a power coefficient curve if you'd like to use FLORIS v3. 

Additionally, it seems that you have a small unit conversion error in your script---power is reported by FLORIS in W, so dividing by `1e3` produces kWh, not MWh.

**Reply by Nellytado** [reply]
Hi, yes that works! Thank you :)
